### PR TITLE
Load scripts and assets from JSON file

### LINKED ISSUE #19: Parity should expose a list of files required for `embed.html` page
Currently we're extracting the files from that page, it would be better if it was exposed directly (e.g. as a JSON file)

### PR DESCRIPTION
**Pull request by ngotchac**
Closes #19

Waiting for https://github.com/ethcore/parity/pull/4426 to get merged.

A `embed.json` file is created from the UI build, with the assets `embed.xxx.js` is dependant on.
Here, we fetch those assets, and replace the URL with the one of the fetched assets.

### CODE REVIEW

File: background/loadScripts.js [hunk 1]
```diff
@@ -43,36 +44,52 @@ export default function loadScripts (port) {
...
+            });
+
+          const scriptPromise = fetch(`http://${UI}/${mainScript}`)
+            .then((response) => response.text());
+
+          return Promise.all([ scriptPromise, Promise.all(assetsPromises) ]);
         })
-        .then(responses => Promise.all(responses.map(response => response.blob())));
+        .then(([ script, assets ]) => {
+          assets.forEach((asset) => {
```
Comment (tomusdrw) [inline background/loadScripts.js]: Why is this required? It seems that it worked fine before, no?

File: background/loadScripts.js [hunk 2]
```diff
@@ -43,36 +44,52 @@ export default function loadScripts (port) {
...
+            script = script.replace(regex, url);
+          });
+
+          return new Blob([ script ], { type: 'application/javascript' });
+        })
+        .then((blob) => {
           return {
             success: true,
-            styles: URL.createObjectURL(styles),
             scripts: URL.createObjectURL(blob)
```
Comment (tomusdrw) [inline background/loadScripts.js]: So now we don't produce separate CSS files even in production?

**Review by tomusdrw — COMMENTED** [review]
Couple of questions.

File: background/loadScripts.js (hunk 1 above)
Comment (ngotchac) [reply background/loadScripts.js]: Well, wasn't this the whole purpose of the issue ? Load the scripts and assets from a JSON file ?

Comment (ngotchac) [reply background/loadScripts.js]: (I forgot to mention the issue : #19)

File: background/loadScripts.js (hunk 2 above)
Comment (ngotchac) [reply background/loadScripts.js]: So it could be both ways. On the UI side, the setup would be that for the embed part, no CSS file is created. For the other part of the UI, CSS files are created, this doesn't change. The idea would be that instead of fetching two files and injecting the CSS manually, we fetch one script that would eventually inject the CSS.

Separate CSS files are great for avoiding FOUC. Here, I think it's not really needed

File: background/loadScripts.js (hunk 1 above)
Comment (tomusdrw) [reply background/loadScripts.js]: Yes, but it seems that here we're replacing some strings inside script content, and I'm wondering why it's required.

Comment (ngotchac) [reply background/loadScripts.js]: Ah, this. Well right now there aren't any really useful assets (no images, etc.), but the idea is that the script might load some images/fonts/etc... If we don't replace the URLs of theses assets, it will try to load, let's say, `/image.png`. This doesn't exist in the Chrome Extension background environment, so it will fail. The best approach IMO would have been to be able to catch these requests and answer with the right URL or already fetched asset. However, Chrome Extensions can't have Service Worker, as you explained, and it cannot catch and modify requests made from within the extension.

So this is the only solution left that I found :/ Not the best though, I agree.

Comment (tomusdrw) [reply background/loadScripts.js]: Hmm, I think it still won't work, cause background script is fetching all resources in `embed.json`, and then transfers those resources to embed.html context. Now in that context we iterate over resources and we just put the `script` tag, so images won't work anyway.

I think I would rather find a way to ensure that embed.js is not referencing any external assets (like images/fonts/etc) and either inline everything or provide from outside.

Comment (ngotchac) [reply background/loadScripts.js]: It's true, I actually thought of that, but was easier this way. Would actually be better. However I dont see why it wouldn't work? It replaces `image.src = "/ethereum.png"` with `image.src = "blob:chrome-extension://<extention-id>/<hash-url>"`, no ? So it could be loaded from anywhere within the Chrome context

Comment (tomusdrw) [reply background/loadScripts.js]: Ah, you're right! Yes, it seems it will work fine.

**Review by tomusdrw — APPROVED** [review]

**Comment by ngotchac** [comment]
Fallback to old version : get the scripts and files from the HTML file if the `embed.json` file is not available.

**Review by tomusdrw — APPROVED** [review]
Perfect!
